packages/arb-avm-cpp/avm_values/include/avm_values/code.hpp

```diff
@@ -235,4 +349,110 @@ class Code {
     }
 };
 
+class RunningCode : public CodeBase<RunningCode>, public Code {
+    friend CodeBase<RunningCode>;
+
+    mutable std::mutex mutex;
+    uint64_t first_segment;
+    std::vector<std::shared_ptr<CodeSegment>> segment_list;
+
+    std::shared_ptr<Code> parent;
+
+    const std::shared_ptr<CodeSegment>& getSegment(uint64_t segment_num) const {
+        return segment_list.at(segment_num - first_segment);
```

Comment:
If this forwards to parent when `segment_num < first_segment`, we could remove special casing in loadCodeSegment and loadCodePoint.

Reply:
Currently the way this is setup, there's a mutex on every level, and getSegment doesn't acquire the mutex so it can't be chained up as is. The way this is setup currently, it's only used for the common implementation stuff in CodeBase to try and minimize duplication

There's not a great scheme for making the mutex and general safety properties of the code self-documenting in the current setup. I'm think about whether there'd be a way to reorganize it to make it more clear

Reply:
Did some refactoring to make this more clear